Add subtype labels

erbridge opened the pull request:
Add labels like the server designation of Patron, but for the chosen subtype for Chameleon and Chimera.

Saintis commented on src/css/base.styl:
```diff
@@ -1528,6 +1528,10 @@ nav ul
               transform(rotate(90deg))
 
+            .subtype-target
+              transform(rotate(90deg), translateX(100%))
```
This rotates it the wrong way for the runner... missing the corresponding code under `.corp-board.opponent` to rotate / translate it back. `transform(rotate(-90deg), translateY(-100%))` seems to do it for me.

Comparison:
Old
<img width="289" alt="screen shot 2016-09-23 at 19 47 19" src="https://cloud.githubusercontent.com/assets/13198563/18796025/b4a60ed8-81c7-11e6-91f0-a38673d9a3e1.png">
New
<img width="288" alt="screen shot 2016-09-23 at 19 54 03" src="https://cloud.githubusercontent.com/assets/13198563/18796036/b9c80a24-81c7-11e6-945f-84d2a095fa10.png">

erbridge replied:
Aha. New to the codebase, so missed that. I'll fix that.

erbridge replied:
Should be fixed now.

Saintis commented:
Some pics to show the awsomeness:
Corp-side:
<img width="283" alt="screen shot 2016-09-23 at 19 45 59" src="https://cloud.githubusercontent.com/assets/13198563/18796506/eec8bb22-81c9-11e6-9293-fbcb866ebec6.png">


Chameleon:
<img width="215" alt="screen shot 2016-09-23 at 19 47 57" src="https://cloud.githubusercontent.com/assets/13198563/18796501/e54108c0-81c9-11e6-9d8b-35a504cb1d4e.png">

Saintis commented:
I do have another discussion point: colours?

Would love to have the following colour scheme:
- Barrier ~ Anarch orange
- Code Gate ~ Shaper green
- Sentry ~ Criminal blue

Would make the icons a bit more fun than the current "sienna" colour 😄

Thoughts?

Saintis commented:
Seems like you have to update the tests to look for `:subtype-target` rather than `:ice-type` for the morph ice.

erbridge commented:
Setting the colours based on the type looks like it needs a better grasp of Clojure than I have (this is the first time I've even looked at the language :)). Doesn't feel like a necessary feature to me, though, but feel free to correct me :).

Looking at the failing test, now. It doesn't refer to `:ice-type` anywhere in the code, and that change was related to Chimera, not the moprh ice anyway. It looks like `update-ice-strength` isn't being fired, but I don't understand why that would be...

Saintis commented:
I'll have a look.

Saintis commented on src/clj/game/cards-ice.clj:
```diff
@@ -336,9 +337,10 @@
                                       (->> (vec (.split (:subtype card) " - "))
                                            (concat [target])
                                            (join " - "))))
+                      (update! (assoc card :subtype-target target))
```
Found the problem. The `update!` changes the card as stored in `state` (signified by the `!`).
Now above this line the card is also `update!`ed to have the new `:subtype`. The `:subtype-target` `update!` overwrites this change (as you are `assoc`ing with the old `card`), which is why the tests fail (the card never actually morphs).

This works for me:
```clojure
(defn morph [state side card new old]
  (update! state side (assoc card
                        :subtype-target new
                        :subtype (->> (remove #(= old %) (.split (:subtype card) " - "))
                                      vec (concat [new]) distinct (join " - "))))
  (update-ice-strength state side card))
```

erbridge replied:
Aha! This makes sense. Is this why I see `(get-card state card)` all over the place?

Saintis replied:
So `(get-card state card)` refreshes the card as defined in the state (so that could have been an alternative solution for the problem above). It's been added in a lot of places recently to prevent client side hacking of trash / rez / play costs or other devious things by reseting the card definition to what the server has. Except for client-facing functions there is normally not much need to call `(get-card state card)` since its often clearer to use a `let` or similar to define the "updated" card before you send it to `update!` or other similar function.

@nealterrell anything to add to this?

erbridge replied:
That's what I thought. That's useful. Just testing the fix now.

nealterrell replied:
Not much to add. Yes, be careful with multiple `update!` to the same card. Values are immutable so the first does not change your local copy of the card data.

erbridge commented:
So I have noticed that if I `/counter 0` a Wendigo with an odd number of advancement tokens, it retains the Barrier type. Is there an event I need to be listening to for this, or is this an issue with `/counter`?

Saintis commented:
Appears to be an issues with `/counter 0`. I don't think it triggers `:advancement-placed`. Feel free to make an issue 😄

erbridge commented:
Alright :).

I think all the issues are fixed (other than the colours). I thought about giving Mother Goddess the same treatment, but she can have a lot of subtypes, so it wouldn't really work in this way. Also Panchatantra, but that's not even implemented.

erbridge commented on src/cljs/netrunner/gameboard.cljs:
```diff
@@ -512,7 +512,12 @@
+          (let [colour-type (case subtype-target
+                              ("Barrier" "Sentry") (lower-case subtype-target)
+                              "Code Gate" "code-gate"
+                              nil)]
```
This was the bit I struggled with :).

Saintis replied:
Feel free to get rid of it if you want. I wanted to see if a PR to your branch worked and this happened instead 😕 😄

erbridge replied:
Pretty cool, though. Maybe update it with the new colours, though?

Saintis commented:
Mother Goddess + Panchatantra would both require some way to have more than one subtype label, which would be cool but probably run into space issues.

Also Tinkering which adds Barrier, Code Gate and Sentry.

Some quick colours:
<img width="281" alt="screen shot 2016-09-23 at 22 13 57" src="https://cloud.githubusercontent.com/assets/13198563/18800153/bf63696a-81db-11e6-822f-32ae9940490d.png">
<img width="209" alt="screen shot 2016-09-23 at 22 14 03" src="https://cloud.githubusercontent.com/assets/13198563/18800155/bfb030ec-81db-11e6-9cac-844ee2b31fcd.png">

Saintis commented:
@mtgred this is ready for merge now. Feel free to skip / revert the colours (or change them) depending on your personal preference.

erbridge commented:
That white on green is pretty low contrast... Might be better to mkae the colours a bit more muted, or change the text colour.

Saintis commented:
Alternative darker colours:
<img width="213" alt="screen shot 2016-09-23 at 22 43 29" src="https://cloud.githubusercontent.com/assets/13198563/18800794/3c2cde42-81df-11e6-8edb-563d89908aad.png">

```stylus
&.barrier
  background-color: #cc5200

&.code-gate
  background-color: #29a329

&.sentry
  background-color: #007acc
```

Saintis commented:
Pushed the change to the darker colours.

Saintis commented:
If we want to add more labels there is a fairly comprehensive list of related cards in #1388.

erbridge commented:
Added Mother Goddess and refactored a bunch of stuff. She gets the first character of subtypes if she has more than one (to prevent wrapping issues).

Also refactored a lot of repeated subtype modification code.

The cards that modify a card other than themselves are a little problematic, due to potential clashes. Probably subtype targets should be given a similar treatment as Rebranding Team gives to the Avertisement subtype.

JoelCFC25 commented:
Really just fantastic work--it looks SO much cleaner.

Saintis commented on src/clj/game/cards-agendas.clj:
```diff
@@ -612,36 +612,27 @@
+         add-ad (fn [state side c]
+                  (update! state side
+                           (assoc c :subtype (combine-subtypes false ;append ad even if it is already an ad
+                                                               (or (:subtype c) "")
```
This could be `(:subtype c "")`

Saintis commented on src/clj/game/cards-agendas.clj:
```diff
@@ -612,36 +612,27 @@
-                                                    (->> (vec (.split (or (:subtype c) "") " - "))
-                                                         (drop 1) ;so that all actual ads remain ads if agenda leaves play
-                                                         (join " - "))))))}
+   (let [get-assets (fn [state corp]
```
Could alternatively use `letfn`

mtgred commented:
Look good. :)
